File: odincal/oops/oops/odintime.py
#!/usr/bin/python

import sys
import math
import time
import string


def fitOrbit(stw):
    c = [-9.631225714726023e-27,
         4.743802284530715e-17,
         1.071801519797240e-05
         - 1.818550814109342e+00]
    o = c[3] + (c[2] + (c[1] + c[0] * stw) * stw) * stw
    return o


def fitMJD(stw):
    c = [1.978082016291609e-31,
         -1.368706688851376e-21,
         7.233098097480957e-07,
         5.196017912410997e+04]
    mjd = c[3] + (c[2] + (c[1] + c[0] * stw) * stw) * stw
    return mjd


class TimeConverter:
    def __init__(self):
        # file = "/archive/time/eqx.mat"
        file = "/home/bengt/test/eqx.mat"
        input = open(file, 'r')
        lines = input.readlines()
        i = 0
        self.orbits = []
        for line in lines:
            cols = string.split(line)
            orbit = int(float(cols[0]))
            JD = float(cols[3])
            if i == orbit:
                self.orbits.append(JD)
                i = i + 1
            else:
                print "index mismatch"
            input.close()

        # file = "/archive/time/STW_PLN.DAT"
        file = "/home/bengt/Downloads/stw_pln.dat"
        input = open(file, 'r')
        lines = input.readlines()
        self.STW = []
        for line in lines:
            cols = string.split(line)
            if len(cols) == 14 and cols[1] == "0000":
                expr = "0x" + cols[7] + "L"
                stw = long(expr, 16)
                if stw < 0:
                    stw = stw + 0x100000000
                # stw = cols[7]
                year = int(float(cols[8]) / 10000.0)
                mon = int(float(cols[8]) / 100.0) % 100
                day = int(float(cols[8])) % 100
                utc = string.split(cols[10], ':')
                hour = int(utc[0])
                min = int(utc[1])
                secs = float(utc[2])
                JD = self.djl(year, mon, day, hour, min, secs)
                rate = float(cols[11])
                self.STW.append([stw, JD, rate])
        input.close()

    def djl(self, year, mon, day, hour, min, secs):
        dn = 367 * year - 7 * (year + (mon + 9) / 12) / 4 \
            - 3 * ((year + (mon - 9) / 7) / 100 + 1) / 4 + 275 * mon / 9 \
            + day + 1721029
        jd = float(dn) - 0.5 + (hour + (min + secs / 60.0) / 60.0) / 24.0
        return jd

    def cld(self, JD):
        (z, f) = divmod(JD + 0.5, 1.0)
        if z < 2299161:
            a = z
        else:
            alpha = int((z - 1867216.25) / 36524.25)
            a = z + 1 + alpha - alpha / 4

        b = a + 1524
        c = int((b - 122.1) / 365.25)
        d = int(365.25 * c)
        e = int((b - d) / 30.6001)

        day = int(b - d - int(30.6001 * e))

        if e < 14:
            month = e - 1
        else:
            month = e - 13

        if month > 2:
            year = c - 4716
        else:
            year = c - 4715

        (z, f) = divmod(24.0 * f, 1.0)
        hour = int(z)
        (z, f) = divmod(60.0 * f, 1.0)
        min = int(z)
        secs = 60.0 * f
        return (year, month, day, hour, min, secs)

    def orbit2JD(self, orbit):
        if orbit < 0.0:
            return 0.0
        i = int(orbit)
        if len(self.orbits) > i + 1:
            r = orbit - i
            JD0 = self.orbits[i]
            JD1 = self.orbits[i + 1]
            JD = JD0 + r * (JD1 - JD0)
        else:
            JD0 = self.orbits[-2]
            JD1 = self.orbits[-1]
            r = orbit - len(self.orbits) + 1.0
            JD = JD1 + r * (JD1 - JD0)
        return JD

    def JD2orbit(self, JD):
        if JD < 2451960.8604086721:
            return 0.0
        for i in range(1, len(self.orbits)):
            if self.orbits[i] > JD:
                r = (JD - self.orbits[i - 1]) / \
                    (self.orbits[i] - self.orbits[i - 1])
                orbit = float(i - 1) + r
                break
        else:
            r = (JD - self.orbits[-2]) / (self.orbits[-1] - self.orbits[-2])
            orbit = float(len(self.orbits)) + r - 2.0
        return orbit

    def stw2JD(self, stw):
        n0 = 0
        n1 = len(self.STW) - 1
        if stw <= self.STW[n0][0]:
            stw0, JD0, rate = self.STW[n0]
        elif stw >= self.STW[n1][0]:
            stw0, JD0, rate = self.STW[n1]
        else:
            while (n1 - n0) > 1:
                n2 = (n1 + n0) / 2
                stw0 = self.STW[n2][0]
                if stw <= stw0:
                    n1 = n2
                else:
                    n0 = n2
            stw0, JD0, rate = self.STW[n0]
        JD = JD0 + (stw - stw0) * rate / 86400.0
        return JD

    def JD2stw(self, JD):
        n0 = 0
        n1 = len(self.STW) - 1
        if JD <= self.STW[n0][1]:
            stw0, JD0, rate = self.STW[n0]
        elif JD >= self.STW[n1][1]:
            stw0, JD0, rate = self.STW[n1]
        else:
            while (n1 - n0) > 1:
                n2 = (n1 + n0) / 2
                JD0 = self.STW[n2][1]
                if JD <= JD0:
                    n1 = n2
                else:
                    n0 = n2
            stw0, JD0, rate = self.STW[n0]
        stw = long(stw0 + math.floor((JD - JD0) * 86400.0 / rate + 0.5))
        return stw

    def timestamp(self, stw):
        JD = self.stw2JD(stw)
        orbit = self.JD2orbit(JD)
        (year, mon, day, hour, min, secs) = self.cld(JD)
        msg = " %08X %10d %10.3f %15.7f %4d-%02d-%02d %02d:%02d:%06.3f" % \
              (stw, stw, orbit, JD, year, mon, day, hour, min, secs)
        return msg


def usage():
    print "usage: %s            " % (sys.argv[0])
    print "       %s -stw number" % (sys.argv[0])
    print "       %s -orbit number" % (sys.argv[0])
    print "       %s -mjd number" % (sys.argv[0])
    print "       %s -jd number" % (sys.argv[0])
    print "       %s -date yyyy-mm-dd hh:mm:ss" % (sys.argv[0])
    sys.exit()


if __name__ == "__main__":
    tc = TimeConverter()
    if len(sys.argv) == 1:
        (year, mon, day, hour, min, secs, wday, yday, isdst) = time.gmtime()
        JD = tc.djl(year, mon, day, hour, min, secs)
        stw = tc.JD2stw(JD)
        orbit = tc.JD2orbit(JD)
#        for orbit in range(400,len(tc.orbits)):
#            JD = tc.orbit2JD(orbit)
#            (year, mon, day, hour, min, secs) = tc.cld(JD)
#            # orbit = tc.JD2orbit(JD)
#            # print "back  %8.2f" % orbit

#            stw = tc.JD2stw(JD)
#            print "%6.0f %15.7f %4d-%02d-%02d %02d:%02d:%06.3f %08X %10d" % \
#                  (orbit, JD, year, mon, day, hour, min, secs, stw, stw)
    else:
        if len(sys.argv) < 3:
            usage()
        if sys.argv[1] == '-stw':
            stw = sys.argv[2]
            if string.find(stw, '0x') == 0:
                stw = long(stw, 16)
            else:
                stw = long(stw)
            JD = tc.stw2JD(stw)
            orbit = tc.JD2orbit(JD)
            (year, mon, day, hour, min, secs) = tc.cld(JD)
        elif sys.argv[1] == '-orbit':
            orbit = float(sys.argv[2])
            JD = tc.orbit2JD(orbit)
            stw = tc.JD2stw(JD)
            (year, mon, day, hour, min, secs) = tc.cld(JD)
        elif sys.argv[1] == '-mjd':
            mjd = float(sys.argv[2])
            JD = mjd + 2400000.5
            stw = tc.JD2stw(JD)
            orbit = tc.JD2orbit(JD)
            (year, mon, day, hour, min, secs) = tc.cld(JD)
        elif sys.argv[1] == '-jd':
            JD = float(sys.argv[2])
            stw = tc.JD2stw(JD)
            orbit = tc.JD2orbit(JD)
            (year, mon, day, hour, min, secs) = tc.cld(JD)
        elif sys.argv[1] == '-date':
            date = sys.argv[2]
            if len(sys.argv) == 3:
                utc = "00:00:00"
            else:
                utc = sys.argv[3]
            (year, mon, day) = string.split(date, '-')
            year = int(year)
            mon = int(mon)
            day = int(day)
            (hour, min, sec) = string.split(utc, ':')
            hour = int(hour)
            min = int(min)
            secs = int(sec)
            JD = tc.djl(year, mon, day, hour, min, secs)
            stw = tc.JD2stw(JD)
            orbit = tc.JD2orbit(JD)
        else:
            usage()

    print " orbit  |   Julian Date   |    date    |     utc      |",
    print "STW(hex) | STW(dec)"
    print string.join(map((lambda x: '-' * x), [8, 17, 12, 14, 10, 11]), '+')
    print "%7.1f | %15.7f |" % (orbit, JD),
    print "%4d-%02d-%02d |" % (year, mon, day),
    print "%02d:%02d:%06.3f |" % (hour, min, secs),
    print "%08X |%10d" % (stw, stw)

File: odincal/oops/Library/odintime.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>
#include <string.h>
#include <sys/stat.h>
#include <unistd.h>

#include "astrolib.h"
#include "odintime.h"

static char rcsid[] = "$Id$";

static struct {
  unsigned long STW0, CLK0;
} STWreset[] = {
  /* { 0x01cf3600, 944660100L }, STW on 1999-12-08 13:35 UTC.           0 */
  /* { 0x0113e400, 947081460L }, STW on 2000-01-05 14:11 UTC.           1 */
  /* { 0x00000000, 950787903L }, STW on 2000-01-17 11:45 UTC.           2 */
  /* { 0x0091ca2a, 962965817L }, STW on 2000-07-07 10:30:17 (Toulouse)  3 */
  /* { 0x030d9733, 975690082L }, STW on 2000-12-01 17:01:22 (Linkoping) 4 */
  /* { 0x004d2e5f, 980815386L }, STW on 2001-01-30 00:43:06 (Svobodny)  5 */
  /* { 0x0014bd34, 981591480L }, STW on 2001-02-08 00:18:00 (Svobodny)  6 */
  { 0x001aeecc, 982752996 }, /* STW on 2001-02-21 10:56:36.90 (Flight)  0 */
};

static unsigned long launch = 2451961L;   /* JD on launch date 12:00 UTC   */

/*static char stw_pln[] = "/archive/time/STW_PLN.DAT";*/
static char stw_pln[] = "/home/bengt/Downloads/stw_pln.dat";
/*static char eqx_mat[] = "/archive/time/eqx.mat";*/
static char eqx_mat[] = "/home/bengt/test/eqx.mat";


struct STWtable {
  unsigned long stw;
  double JD;
  double rate;
};

static int stwnum, eqxnum;
static struct STWtable *stwtbl;
static double *eqx;

static void readEQX_MAT()
{
  FILE *f;
  char *ptr;
  static char line[256];
  int i, n, need;
  double JD;
  int orbit;
  struct stat buf;
 
  f = fopen(eqx_mat, "r");
  if (f == NULL) {
    fprintf(stderr, "can't open '%s'\n", eqx_mat);
    exit(1);
  }
  

  if (stat(eqx_mat, &buf) == -1) {
    fprintf(stderr, "can't stat '%s'\n", eqx_mat);
    exit(1);
  }

  need = (int)buf.st_size;
  
  need = need/101;
  eqx = NULL;
  if ((eqx = calloc(need, sizeof(double))) == NULL) {
    fprintf(stderr, "memory allocation failure\n");
    exit(1);
  }
  eqxnum = need;
  for (orbit = 0; orbit < eqxnum; orbit++) {
    eqx[orbit] = 0.0;
  }
  
  n = 0;
  while (fgets(line, 255, f) != NULL) {
    i = strlen(line);
    if ((ptr = rindex(line, '\n')) != NULL) *ptr = '\0';
    if ((ptr = rindex(line, '\r')) != NULL) *ptr = '\0';
    ptr = strtok(line, " \t\n\0");
    i = 0;
    orbit = -1;
    JD = 0.0;
    while (ptr) {
      if (i == 0) {
	orbit = (int)floor(atof(ptr));
      }
      if (i == 3) {
	JD = atof(ptr);
      }
      ptr = strtok(NULL, " \t\n\0");
      i++;
    }
    if ((orbit < 0) || (orbit >= eqxnum) || (JD == 0.0)) {
      fprintf(stderr, "orbit table failure\n");
      exit(1);
    }
    eqx[orbit] = JD;
  }
  fclose(f);
}

static void readSTW_PLN()
{
  FILE *f;
  char *ptr;
  static char line[256];
  int i, n, need;
  unsigned long stw, date;
  int year, month, day;
  double secs;
  double rate;
  struct stat buf;

  f = fopen(stw_pln, "r");
  if (f == NULL) {
    fprintf(stderr, "can't open '%s'\n", stw_pln);
    exit(1);
  }
  
  if (stat(stw_pln, &buf) == -1) {
    fprintf(stderr, "can't stat '%s'\n", stw_pln);
    exit(1);
  }

  need = (int)buf.st_size;
  need = need/153;
  if ((stwtbl = calloc(need, sizeof(struct STWtable))) == NULL) {
    fprintf(stderr, "memory allocation failure\n");
    exit(1);
  }
  stwnum = need;

  n = 0;
  while (fgets(line, 255, f) != NULL) {
    if (line[0] == '!') continue;
    i = strlen(line);
    if ((ptr = rindex(line, '\n')) != NULL) *ptr = '\0';
    if ((ptr = rindex(line, '\r')) != NULL) *ptr = '\0';
    ptr = strtok(line, " \t\n\0");
    i = 0;
    while (ptr) {
      if ((i == 1) && strncmp(ptr, "0000", 4)) break;
      if (i == 7) {
	stw = strtoul(ptr, NULL, 16);
	stwtbl[n].stw = stw;
	stwtbl[n+1].stw = 0L;
      }
      if (i == 8) {
	date = strtoul(ptr, NULL, 10);
	year = date/10000;
	month = (date/100) % 100;
	day = date % 100;
	stwtbl[n].JD = djl(year, month, day);
      }
      if (i == 9) {
	secs = atof(ptr);
	stwtbl[n].JD += secs/86400.0;
      }
      if (i == 11) {
	rate = atof(ptr);
	stwtbl[n].rate = rate;
	n++;
	if (n == stwnum) {
	  fprintf(stderr, "table size exceeded\n");
	  exit(1);
	}
      }
      ptr = strtok(NULL, " \t\n\0");
      i++;
    }
  }
  stwnum = n;
  fclose(f);
}

static double orbit2JD(double orbit)
{
  double r, o, JD;
  int i;

  if (orbit < 0.0) return 0.0;
  if (eqxnum == 0) readEQX_MAT();
  r = modf(orbit, &o);
  i = (int)o;
  if (i < eqxnum-1) {
    JD = eqx[i]+r*(eqx[i+1]-eqx[i]);
  } else {
    r = orbit-(double)eqxnum+1.0;
    JD = eqx[eqxnum-1]+r*(eqx[eqxnum-1]-eqx[eqxnum-2]);
  }

  return JD;
}

static double JD2orbit(double JD)
{
  int i;
  double orbit, r;

  if (JD < 2451960.8604086721) return 0.0;
  if (eqxnum == 0) readEQX_MAT();

  for (i = 1; i < eqxnum; i++) {
    if (eqx[i] >= JD) {
      r = (JD-eqx[i-1])/(eqx[i]-eqx[i-1]);
      orbit = (double)(i-1)+r;
      break;
    }
  }
  if (i == eqxnum) {
    r = (JD-eqx[eqxnum-2])/(eqx[eqxnum-1]-eqx[eqxnum-2]);
    orbit = (double)eqxnum+r-2.0;
  }
  return orbit;
}

static double stw2JD(unsigned long stw)
{
  int n0, n1, n2;
  unsigned long stw0;
  double rate, JD, JD0;

  if (stwnum == 0) readSTW_PLN();

  n0 = 0;
  n1 = stwnum-1;
  if (stw <= stwtbl[n0].stw) {
    stw0 = stwtbl[n0].stw;
    JD0  = stwtbl[n0].JD;
    rate = stwtbl[n0].rate;
  } else if (stw >= stwtbl[n1].stw) {
    stw0 = stwtbl[n1].stw;
    JD0  = stwtbl[n1].JD;
    rate = stwtbl[n1].rate;
  } else {
    while ((n1-n0) > 1) {
      n2 = (n1+n0)/2;
      stw0 = stwtbl[n2].stw;
      if (stw <= stw0) n1 = n2;
      else             n0 = n2;
    }
    stw0 = stwtbl[n0].stw;
    JD0  = stwtbl[n0].JD;
    rate = stwtbl[n0].rate;
  }
  if (stw >= stw0) JD = JD0 + (stw-stw0)*rate/86400.0;
  else             JD = JD0 - (stw0-stw)*rate/86400.0;

  return JD;
}

static unsigned long JD2stw(double JD)
{
  int n0, n1, n2;
  double rate, JD0;
  unsigned long stw, stw0;

  if (stwnum == 0) readSTW_PLN();
  n0 = 0;
  n1 = stwnum-1;
  if (JD <= stwtbl[n0].JD) {
    stw0 = stwtbl[n0].stw;
    JD0  = stwtbl[n0].JD;
    rate = stwtbl[n0].rate;
  } else if (JD >= stwtbl[n1].JD) {
    stw0 = stwtbl[n1].stw;
    JD0  = stwtbl[n1].JD;
    rate = stwtbl[n1].rate;
  } else {
    while ((n1-n0) > 1) {
      n2 = (n1+n0)/2;
      JD0 = stwtbl[n2].JD;
      if (JD <= JD0) n1 = n2;
      else           n0 = n2;
    }
    stw0 = stwtbl[n0].stw;
    JD0  = stwtbl[n0].JD;
    rate = stwtbl[n0].rate;
  }
  stw = stw0 + (long)floor((JD-JD0)*86400.0/rate+0.5);
  return stw;
}

static int validreset(int reset)
{
  int n;
  
  n = sizeof(STWreset)/sizeof(STWreset[0]);
  if (reset < 0 || reset >= n) reset = n-1;
  
  return reset;
}

/*
  Calculate modified Julian date from given date and time.
*/
double mjd(int year, int month, int day, int hour, int min, double secs)
{
  double JD, mjd;

  JD = djl(year, month, day);
  secs += (double)(hour*3600 + min*60);
  JD += secs/86400.0;
  mjd = jd2mjd(JD);

  return mjd;
}

/* 
   The following routine converts STW to mjd. 
 */
double stw2mjd(unsigned long stw, int reset)
{
  double JD, mjd;

  JD = stw2JD(stw);
  mjd = jd2mjd(JD);

  return mjd;
}

/* 
   The following routine converts mjd to STW. 
 */
unsigned long mjd2stw(double mjd, int reset)
{
  double JD;
  unsigned long stw;

  JD = mjd2jd(mjd);
  stw = JD2stw(JD);

  return stw;
}

/* 
   The following routine converts STW to UTC. 
   It returns number of seconds since 1970-01-01.
 */
unsigned long stw2utc(unsigned long STW, int reset)
{
  unsigned long utc; 
  double mjd;

  mjd = stw2mjd(STW, reset);
  utc = mjd2utc(mjd);

  return utc;
}

/* 
 * The following routine converts UTC to STW. 
 * It takes number of seconds since 1970-01-01.
 */
unsigned long utc2stw(unsigned long utc, int reset)
{
  unsigned long STW;
  double mjd;

  mjd = utc2mjd(utc);
  STW = mjd2stw(mjd, reset);

  return STW;
}

/* 
 * The following routine converts UTC to MJD. 
 * It takes number of seconds since 1970-01-01.
 */
double utc2mjd(unsigned long utc)
{
  struct tm *gmt;
  double secs;
  int year, month, day, hour, min;

  gmt = gmtime(&utc);
  year  = gmt->tm_year+1900;
  month = gmt->tm_mon+1;
  day   = gmt->tm_mday;
  hour  = gmt->tm_hour;
  min   = gmt->tm_min;
  secs  = (double)gmt->tm_sec;
  
  return mjd(year, month, day, hour, min, secs);
}

/* 
 * Inverse of the above.
 */
unsigned long mjd2utc(double mjd)
{
  struct tm gmt;
  unsigned long now;
  double secs, JD;

  JD = mjd2jd(mjd);
  cld(JD, &gmt.tm_year,&gmt.tm_mon,&gmt.tm_mday,
      &gmt.tm_hour,&gmt.tm_min,&secs);
  gmt.tm_sec = (int)floor(secs);
  gmt.tm_year -= 1900;
  gmt.tm_mon  -= 1;
  gmt.tm_wday = 0;
  gmt.tm_yday = 0;
  
  gmt.tm_isdst = 0;
  now = mktime(&gmt)+3600;

  return now;
}

/* 
 * The following routine converts MJD to orbit number.
 * Will return 0.0 for days before 2001-02-20.
 */
double mjd2orbit(double mjd)
{
  double orbit, JD;

  JD = mjd2jd(mjd);
  orbit = JD2orbit(JD);

  return orbit;
}

/* 
 * Inverse of the above.
 */
double orbit2mjd(double orbit)
{
  double JD, mjd;

  JD = orbit2JD(orbit);
  mjd = jd2mjd(JD);

  return mjd;
}

/* 
 * The following routine converts UTC to orbit number. 
 * It takes number of seconds since 1970-01-01.
 */
double utc2orbit(unsigned long utc)
{
  return mjd2orbit(utc2mjd(utc));
}

/* 
 * Inverse of the above.
 */
unsigned long orbit2utc(double orbit)
{
  return mjd2utc(orbit2mjd(orbit));
}

/* 
 * Produce an ASCII representation of UTC.
 */
char *asciitime(unsigned long utc)
{
  static char zeit[32], *at;
  struct tm *gmt;
  int n;

  gmt = gmtime(&utc);
  at = asctime(gmt);
  strcpy(zeit, at);
  n = strlen(zeit);
  /* remove EOL at end of string */
  zeit[n-1] = '\0';

  return zeit;
}

#ifdef MAIN
int main()
{
    unsigned long stw0 = 0x80000000;
    unsigned long stw1, utc;
    double JD;

    JD = stw2JD(stw0);
    stw1 = JD2stw(JD);
    printf("%10u %15.6f %10u\n", stw0, JD, stw1);

    stw0 = 0x7fffffff;
    utc = stw2utc(stw0, -1);
    stw1 = utc2stw(utc, -1);
    printf("%10u %10d %10u\n", stw0, utc, stw1);

    stw0 = 0x80000000;
    utc = stw2utc(stw0, -1);
    stw1 = utc2stw(utc, -1);
    printf("%10u %10d %10u\n", stw0, utc, stw1);

    exit(0);
}

#endif
